File: js/ai/mcts.js
/* NOTES:
	- This is actually a varient of the MCTS-Solver, (which is stronger than plain MCTS)
	- Scoring is local, but when considering a board state, the turn has already been changed,
	  so, the local node is actually opposite of board.turn
	- A node's score will be in the range of [-1,+1], or = -infinity, +infinity
*/
var MCTS_MAX_ITERATIONS = 100000;
var MCTS_UCT_TUNING = 0.9; //Controls exploration (< 1) vs. exploitation (> 1)
var MCTS_SECURE_TUNING = 1;
var MCTS_VISITS_TO_ADD_NODE = 1;
var MCTS_MIN_FAIR_PLAY = 1;

var MCTS_WIN_SCORE = 1;
var MCTS_LOSE_SCORE = -1;
var MCTS_TIE_SCORE = 0;


//Class MCTS
function MCTS(board) {    
    this.board = board;		
}

MCTS.prototype.getMove = function() {		
	//Check for available wins before trying to build the search tree
	var winFound = this.board.findWin();
	if (winFound) {
		scoreEnabled = false;
		return this.board.getMoveFromMidWin(winFound);	
	}
	//TODO: check for ties?	
	//Run the monte-carlo tree search
	console.log('MCTS: Initial board');
	this.board.print();
	var bestKid = this.runMCTS(this.board.clone());	
	var move = this.board.deriveMove(bestKid.board);
	console.log('Visits: ' + bestKid.visits + ', Score: ' + bestKid.score);		
	this.board.printMove(move);
	bestKid.board.print();
	return move;
}

MCTS.prototype.runMCTS = function(board) {	     
    /* Steps:
	0. Pre-expansion
	---- Steps done each iteration -----
    1. Selection
    2. Expansion
    3. Simulation
    4. Back-propagation        
	-----------------------------------
    5. Pick final move
	*/
    var root = {visits:0, score:0, val:0, board:board, parent:null, kids:[]};
	
	//Pre-expand root's children
	if (!this.preExpand(root)) {
		root.board.makeRandomMove(); //If there are no moves available
        return root; 
	}
	
	//Run the iteration steps
	for (var i = 0; i < MCTS_MAX_ITERATIONS; i++) {	
	
		//Select
        var node = this.selectNode(root);
		
		//Expand - but make sure an adequate number of simulations have been run before expanding.		
		if (node.visits >= MCTS_VISITS_TO_ADD_NODE) {
			var terminalScore = this.expandNode(node); 	
			if (terminalScore !== false) {
				var winFound = this.backpropagate(node, terminalScore);
				if (winFound) return winFound;
			}
		}
        else {
            //Simulate
            var simScore = this.simulate(node);
		
            //Backpropagate			
            var winFound = this.backpropagate(node, simScore);
			if (winFound) return winFound; 
        }
	}	
	//Pick the final move
	return this.pickFinalMove(root);
}

//Steps
MCTS.prototype.preExpand = function(root) {
	var moves = this.board.getAllNonLossMoves();       
    if (moves.length == 0) {
        alert('No non-loss moves found - making random');
        return false;   
    }
	for (var m = 0; m < moves.length; m++) {
        root.kids.push({visits:0, score:0, val:0, board:moves[m], parent:root, kids:[]});
    }
	return true;
}

MCTS.prototype.selectNode = function(root) {
    //Traverse the tree until a leaf is reached by selecting the best UCT	
    var node = root;
    while (node.kids.length > 0) {
		var bestUCT = -INFINITY;
		var bestNode = null;		
		var length = node.kids.length;
		for (var i = 0; i < length; i++ ){ 
			var kid = node.kids[i];		
			//Make sure all nodes get a fair chance
			if (kid.visits < MCTS_MIN_FAIR_PLAY) return kid;
			
			//Upper Confidence Bound for Trees - Multi-armed bandit problem to maximize payoff
			var uct = (kid.score + MCTS_UCT_TUNING) * Math.sqrt(Math.log(kid.parent.visits) / kid.visits); //UCT must be negative?
			if (uct > bestUCT) {
				bestUCT = uct;
				bestNode = kid;
			}
		} 

		node = bestNode;
		if (bestNode == null) {
			console.log('All moves lead to loss'); 			
			return root.kids[0];
		}
    }	
	return bestNode;       
   
}

MCTS.prototype.expandNode = function(node) {	
	var board = node.board;	
    //Check for terminal state - they need backpropagated instead of expanding	    
    var winFound = board.findWin();	
	if (winFound) return -INFINITY; //Negative because it's from the parent's point of view	

    //TODO: Handle dual wins?
    
    //Else get all possible unique moves that don't lead to a loss
    var moves = board.getAllNonLossMoves();
    if (moves.length == 0) {
		if (board.getAllMoves().length) return INFINITY; //All moves lead to loss - So parent's win
		else return MCTS_TIE_SCORE; //Tie - no moves available
		
    }
	
	//Add all children to the node    
	for (var m = 0; m < moves.length; m++) {
		node.kids.push({visits:0, score:0, val: 0, board:moves[m], parent:node, kids:[]});
	}
	return false;
}

MCTS.prototype.simulate = function(node) {
    //Scoring from point of view of node
    var board = node.board.clone();	
    var moveCount = bitCount(or(board.p1, board.p2));
	var curPlayer = !board.turn;
	var movesLeft = BOARD_SPACES - moveCount;
    for (var i = 0; i < movesLeft; i++) {
        //Check for wins
        var winFound = board.findWin();
        if (winFound) {			
            if (board.turn == curPlayer) return MCTS_WIN_SCORE; //Board turn hasn't changed yet
            else return MCTS_LOSE_SCORE;			
			//TODO: test for dual win
		}
        else { //Block opponent win
            board.turn = !board.turn;
            winFound = board.findWin();
            if (winFound) {
                var move = board.getMoveFromMidWin(winFound);
                board.turn = !board.turn;
                board.setPin(move.pos);
                board.rotate(move.quad, move.rot);
            }
            else { //Safe to make a random move
                board.turn = !board.turn;
                board.makeRandomMove();
            }
        }
        
        //Make random moves
        //board.makeRandomMove();           
    }    
    return MCTS_TIE_SCORE;
}

MCTS.prototype.backpropagate = function(node, score) {	
	//Update leaf
    node.visits++;
    node.val += score;
	if (Math.abs(score) == INFINITY) node.score = score; //Don't average score if terminal position			
	else node.score = (node.score + score) / node.visits; //Average    

    //Backpropagate from the leaf's parent up to the root, inverting the score each level due to minmax
    while (node.parent != null) {
        score *= -1;        
        node = node.parent;
        node.visits++;		
        node.val += score;
        //Loss - if a child can win then that is a loss for the parent
        if (score == -INFINITY) node.score = -INFINITY;
        
        //Win - but all children must be checked to prove parent win
        else if (score == INFINITY) {
            var loss = true;
            for (var i = 0; i < node.kids.length; i++) {
                if (node.kids[i].score != -INFINITY) {
                    loss = false;
                    break;
                }
            }
            if (loss) node.score = INFINITY; //Proven win
            else node.score = MCTS_LOSE_SCORE; //Can't prove parent win, so not terminal
        }
        
        //Non-terminal position, so just average score        			
		else node.score = (node.score + score) / node.visits;
    }	
	//Check to see if win has been propagated up in direct decendents of root (i.e. first level)	
	if (score == INFINITY) {
		var kids = node.kids; //node is now equal to root
		for (var k = 0; k < kids.length; k++) { 
			if (kids[k].score == INFINITY) {
				alert ('Win propagated');
				return kids[k];			
			}
		}		
	}
	return false;
}

MCTS.prototype.pickFinalMove = function(root) {
	//Should we use secure child?	
    var bestScore = -INFINITY;
	var bestNode = null;
	
	initScoreMap();
	
	for (var i = 0; i < root.kids.length; i++) {
		var kid = root.kids[i];
        var score = kid.score + (MCTS_SECURE_TUNING / Math.sqrt(kid.visits));		
        if (score > bestScore) {
			bestScore = score;
			bestNode = kid;
		}
		//Populate score map
		var move = root.board.deriveMove(kid.board);
        var r = ROW[move.pos];
        var c = COL[move.pos];       
		scoreMap[r][c].push({visits:kid.visits, score:String(kid.val) + '(' + kid.score.toFixed(4) + ')'});
	}	
	
	if (bestNode == null) return root.kids[Math.floor(Math.rand() * root.kids.length)]; //All moves lead to loss
	else {
		var move = root.board.deriveMove(bestNode.board);
		enableScoreMap(move, root.visits);
		return bestNode;
	}
}


//End class MCTS

File: cpp/tork5/tork5/mcts.cpp
#include <vector>
#include <ctime>
#include "Node.h"
using namespace std;

//Constants
const int MAX_ITERATIONS = 10000;
const float UCT_TUNING = 0.9f;
const int VISITS_TO_ADD_NODE = 1;
const int MIN_FAIR_PLAY = 0;
const int INFINITY = 1000000;

const int WIN_SCORE = 1;
const int LOSE_SCORE = -1;
const int TIE_SCORE = 0;


Node* MCTS_pickFinalMove(Node* root) {
	//Max visits
	int bestVisits = -INFINITY;
	Node* bestNode = NULL;
		
	for (int i = 0; i < root->kids.size(); i++) {
		Node* kid = root->kids[i];
		if (kid->visits > bestVisits) {
			bestVisits = kid->visits;
			bestNode = kid;
		}
	}
	
	if (bestNode == NULL) {
		printf("All moves lead to loss: %s\n", root->board.toString().c_str());		
		return root->kids[rand() % root->kids.size()]; //All moves lead to loss
	}
	else return bestNode;
}

void MCTS_backpropagate(Node* node, float score) {
	//Update leaf
    node->visits++;
	if (abs(score) == INFINITY) node->score = score; //Don't average score if terminal position			
	else node->score = ((node->score * (node->visits - 1)) + score) / node->visits; //Average    

    //Backpropagate from the leaf's parent up to the root, inverting the score each level due to minmax
    while (node->parent != NULL) {		
        score *= -1;
        node = node->parent;
        node->visits++;
        
        //Win
        if (score == INFINITY) node->score = INFINITY;
        
        //Loss - but all children must be checked to prove parent loss
        else if (score == -INFINITY) {
            bool loss = true;
            for (int i = 0; i < node->kids.size(); i++) {
                if (node->kids[i]->score != -INFINITY) {
                    loss = false;
                    break;
                }
            }
            if (loss) node->score = -INFINITY;
            else node->score = LOSE_SCORE; //Can't prove parent loss, so not terminal
        }
        
        //Non-terminal position, so just average score        			
		else node->score = ((node->score * (node->visits - 1)) + score) / node->visits;
    }
}

float MCTS_simulate(Node* node) {
	//Scoring from point of view of node
	Board board = node->board.clone();
	int moveCount = bitCount(Or(board.p1, board.p2));

	for (int i = 0; i < (BOARD_SPACES - moveCount); i++) {
        //Check for wins
        int winFound = board.findWin();
        if (winFound) {
            //TODO: test for dual win
            return WIN_SCORE;
        }
        
        //Make random moves
        board.makeRandomMove();   
        //board.print();
    }    
    return TIE_SCORE;
}

void MCTS_expandNode(Node* node) {
	Board board = node->board;	

    //Check for win - don't expand if we can win	    
    int winFound = board.findWin();
	if (winFound) {
		MCTS_backpropagate(node, -INFINITY); //Negative because it's from the parent's point of view	
	}

	else {

		//TODO: Handle dual wins?
    
		//Else add all possible unique moves that don't lead to a loss
		vector<Board>moves = board.getAllNonLossMoves();
    
		//Tie
		if (moves.size() == 0) {
			return MCTS_backpropagate(node, TIE_SCORE);
		}

		//Add kids
		else {
			for (int i = 0; i < moves.size(); i++) {
				node->kids.push_back(new Node(node, moves[i]));
			}
		}
	}
}

Node* MCTS_selectNode(Node* root) {
	//Traverse the tree until a leaf is reached by selecting the best UCT	
    Node* node = root;	
    while (node->kids.size() > 0) {	
		
		float bestUCT = -INFINITY;
		Node* bestNode = NULL;
		float uct;
		int parentVisits = (node->parent == NULL || node->parent->visits == 0)? 1 : node->parent->visits;
		for (int i = 0; i < node->kids.size(); i++) {
			Node* kid = node->kids[i];		
			//Make sure unvisited nodes get a chance
			if (kid->visits <= MIN_FAIR_PLAY) {
				bestNode = kid;
				break;
			
			}						
			uct =  kid->score + sqrt(abs(UCT_TUNING * log((float)parentVisits / kid->visits))); //UCT must be negative?
			if (uct > bestUCT) {
				bestUCT = uct;
				bestNode = kid;
			}
		} 

		node = bestNode;
		if (bestNode == NULL) {
			printf("No best kid found - broken select! UCT: %f, Board: %s", uct, root->board.toString().c_str());
		}
    }	
	return node;
}

Board MCTS_engine(Board board) {
	//Steps:
	//0. Pre-expansion
    //1. Selection
    //2. Expansion
    //3. Simulation
    //4. Back-propagation        
    //5. Pick final move
	clock_t startTime = clock();	
	Node root(NULL, board);	
	//Pre-expand root's children	
	vector<Board>moves = board.getAllNonLossMoves();    		
	for (int i = 0; i < moves.size(); i++) {
		root.kids.push_back(new Node(&root, moves[i]));
	}
	
	for (int i = 0; i < MAX_ITERATIONS; i++) {
		//TODO: top-level win propagation

		//Select
		Node* node = MCTS_selectNode(&root);

		//Expand
		if (node->visits >= VISITS_TO_ADD_NODE) MCTS_expandNode(node); //Expand backpropagates for terminal nodes
		else {
			//Simulate
			float score = MCTS_simulate(node);

			//Backpropagate
			MCTS_backpropagate(node, score);
		}
	}

	//Final move
	Node* finalNode = MCTS_pickFinalMove(&root);
	Board moveBoard = finalNode->board;
	int visits = finalNode->visits;
	float score = finalNode->score;
	root.Cleanup();

	//Log stats			
	double duration = (clock() - startTime) / (double) CLOCKS_PER_SEC;
	printf("Computer: Visits: %i, Score: %f, Time: %f\n", visits, score, duration);
	
	return moveBoard;
}

void MCTS_getMove(Board board, int& pos, int& quad, int& rot) {	
	int winFound = board.findWin();
	if (winFound) board.getMoveFromMidWin(winFound, pos, quad, rot); 
	else {		
		Board moveBoard = MCTS_engine(board);		
		moveBoard.print();				
		board.deriveMove(moveBoard, pos, quad, rot);						
	}	
}